File: problems/640_solve_the_equation/Solution.java
class Solution {
    public String solveEquation(String equation) {
        String[] parts = equation.split("=");
        int[] left = parse(parts[0]);
        int[] right = parse(parts[1]);
        int coef = left[0] - right[0];
        int val = right[1] - left[1];
        if (coef == 0) return val == 0 ? "Infinite solutions" : "No solution";
        return "x=" + (val / coef);
    }
    private int[] parse(String expr) {
        int coef = 0, val = 0;
        String[] tokens = expr.replace("-", "+-").split("\\+");
        for (String token : tokens) {
            if (token.isEmpty()) continue;
            if (token.contains("x")) {
                String t = token.replace("x", "");
                if (t.isEmpty() || t.equals("+")) coef += 1;
                else if (t.equals("-")) coef -= 1;
                else coef += Integer.parseInt(t);
            } else {
                val += Integer.parseInt(token);
            }
        }
        return new int[]{coef, val};
    }
}

File: problems/640_solve_the_equation/solution.py
class Solution:
    def solveEquation(self, equation: str) -> str:
        import re
        def parse(expr):
            coef = val = 0
            tokens = re.findall(r'[+-]?[^+-]+', expr)
            for token in tokens:
                token = token.strip()
                if 'x' in token:
                    t = token.replace('x', '')
                    if t in ('', '+'): coef += 1
                    elif t == '-': coef -= 1
                    else: coef += int(t)
                else:
                    val += int(token)
            return coef, val
        left, right = equation.split('=')
        lc, lv = parse(left)
        rc, rv = parse(right)
        coef = lc - rc
        val = rv - lv
        if coef == 0:
            return 'Infinite solutions' if val == 0 else 'No solution'
        return 'x=' + str(val // coef)
